File: testing/acceptance/temperature-spa/shared/runner/templates.test.ts
// Unit tests for the bead-template handlers (gm-root.27.2 / gm-root.27.28).
//
// Validates the content registry produces correct files and the
// noop / write-component templates write what their frontmatter
// declares. Heavier templates (npm-install, build, serve) shell out
// to npm/vite and are smoke-tested only via the end-to-end run
// (gm-root.27.29) — node:test isolation makes mocking npm awkward.

import { strict as assert } from 'node:assert';
import { describe, it } from 'node:test';
import { mkdtempSync, readFileSync, rmSync, existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { getTemplate } from './templates.ts';

function tempDir(): { dir: string; cleanup: () => void } {
  const dir = mkdtempSync(join(tmpdir(), 'tspa-templates-'));
  return { dir, cleanup: () => rmSync(dir, { recursive: true, force: true }) };
}

describe('getTemplate', () => {
  it('returns a handler for every declared template', () => {
    for (const name of [
      'init-repo',
      'npm-install',
      'write-component',
      'write-test',
      'build',
      'serve',
      'error-then-recover',
      'noop',
    ]) {
      assert.ok(getTemplate(name), `expected template '${name}'`);
    }
  });

  it('returns undefined for an unknown template', () => {
    assert.equal(getTemplate('not-a-template'), undefined);
    assert.equal(getTemplate(undefined), undefined);
  });
});

describe('noop template', () => {
  it('writes every file declared in frontmatter from the registry', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('noop')!;
      await handler(
        { projectDir: dir, beadID: 'tspa-2', labels: ['milestone:m1'] },
        {
          template: 'noop',
          files: ['tsconfig.json', 'index.html', 'src/main.tsx'],
          extras: {},
        },
      );
      assert.ok(existsSync(join(dir, 'tsconfig.json')));
      assert.ok(existsSync(join(dir, 'index.html')));
      assert.ok(existsSync(join(dir, 'src/main.tsx')));
    } finally {
      cleanup();
    }
  });

  it('throws on a file path with no registry entry', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('noop')!;
      await assert.rejects(
        handler(
          { projectDir: dir, beadID: 'tspa-x', labels: [] },
          {
            template: 'noop',
            files: ['unknown/path.ts'],
            extras: {},
          },
        ),
        /no registry entry/,
      );
    } finally {
      cleanup();
    }
  });
});

describe('write-component template + content registry', () => {
  it('writes M2 App.tsx (Hello world) when labels include milestone:m2', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('write-component')!;
      await handler(
        { projectDir: dir, beadID: 'tspa-4', labels: ['milestone:m2'] },
        {
          template: 'write-component',
          files: ['src/App.tsx'],
          extras: {},
        },
      );
      const body = readFileSync(join(dir, 'src/App.tsx'), 'utf8');
      assert.match(body, /data-testid="app-root"/);
      assert.match(body, /Hello world/);
      assert.doesNotMatch(body, /TemperatureTable/);
    } finally {
      cleanup();
    }
  });

  it('writes M3 App.tsx (TemperatureTable wrapper) when labels include milestone:m3', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('write-component')!;
      await handler(
        { projectDir: dir, beadID: 'tspa-11', labels: ['milestone:m3'] },
        {
          template: 'write-component',
          files: ['src/App.tsx'],
          extras: {},
        },
      );
      const body = readFileSync(join(dir, 'src/App.tsx'), 'utf8');
      assert.match(body, /TemperatureTable/);
      assert.match(body, /data-testid="app-root"/);
    } finally {
      cleanup();
    }
  });

  it('TemperatureTable.tsx renders 16 rows with row-{c} testids', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('write-component')!;
      await handler(
        { projectDir: dir, beadID: 'tspa-9', labels: ['milestone:m3'] },
        {
          template: 'write-component',
          files: ['src/TemperatureTable.tsx'],
          extras: {},
        },
      );
      const body = readFileSync(join(dir, 'src/TemperatureTable.tsx'), 'utf8');
      assert.match(body, /data-testid="temperature-table"/);
      assert.match(body, /data-testid=\{`row-\$\{r\.celsius\}`\}/);
    } finally {
      cleanup();
    }
  });

  it('temperatureRows returns °F = °C × 9/5 + 32 to one decimal', async () => {
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('write-component')!;
      await handler(
        { projectDir: dir, beadID: 'tspa-8', labels: ['milestone:m3'] },
        {
          template: 'write-component',
          files: ['src/temperatureRows.ts'],
          extras: {},
        },
      );
      const body = readFileSync(join(dir, 'src/temperatureRows.ts'), 'utf8');
      // The function should iterate 0..300 step 20 and format to 1 decimal.
      assert.match(body, /c <= 300/);
      assert.match(body, /c \+= 20/);
      assert.match(body, /9.+5.+32/);
      assert.match(body, /toFixed\(1\)/);
    } finally {
      cleanup();
    }
  });
});

describe('error-then-recover template', () => {
  it('throws on first attempt, succeeds thereafter', async () => {
    // The module-level retry counter persists for the test process,
    // so we just verify "two consecutive calls — first throws,
    // second succeeds." Subsequent test files that exercise the
    // same template would observe persistent state; the acceptance
    // run only uses error-then-recover once per bead so this is OK.
    const { dir, cleanup } = tempDir();
    try {
      const handler = getTemplate('error-then-recover')!;
      const ctx = { projectDir: dir, beadID: 'tspa-x', labels: [] };
      const fm = { extras: {} };
      let firstThrew = false;
      try {
        await handler(ctx, fm);
      } catch (e) {
        firstThrew = (e as Error).message.includes('deliberate first-attempt failure');
      }
      assert.ok(firstThrew, 'first attempt must throw');
      // Second call must succeed.
      await handler(ctx, fm);
    } finally {
      cleanup();
    }
  });
});

File: internal/adapter/mock/templates_test.go
package mock

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func tempDir(t *testing.T) string {
	t.Helper()
	d, err := os.MkdirTemp("", "mock-tpl-")
	if err != nil {
		t.Fatalf("mkdtemp: %v", err)
	}
	t.Cleanup(func() { os.RemoveAll(d) })
	return d
}

func TestGetTemplate_AllNames(t *testing.T) {
	for _, name := range []string{
		"init-repo", "npm-install", "write-component", "write-test",
		"build", "serve", "error-then-recover", "noop",
	} {
		if GetTemplate(name) == nil {
			t.Errorf("expected template %q", name)
		}
	}
}

func TestGetTemplate_Unknown(t *testing.T) {
	if GetTemplate("not-a-template") != nil {
		t.Error("expected nil for unknown template")
	}
}

func TestNoop_WritesFiles(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("noop")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-2", Labels: []string{"milestone:m1"}},
		Frontmatter{Files: []string{"tsconfig.json", "index.html", "src/main.tsx"}, Extras: map[string]string{}},
	)
	if err != nil {
		t.Fatalf("noop: %v", err)
	}
	for _, rel := range []string{"tsconfig.json", "index.html", "src/main.tsx"} {
		if _, err := os.Stat(filepath.Join(dir, rel)); err != nil {
			t.Errorf("expected %q present: %v", rel, err)
		}
	}
}

func TestNoop_UnknownFileFails(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("noop")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-x", Labels: []string{}},
		Frontmatter{Files: []string{"unknown/path.ts"}, Extras: map[string]string{}},
	)
	if err == nil || !strings.Contains(err.Error(), "no registry entry") {
		t.Fatalf("expected no-registry-entry error; got %v", err)
	}
}

func TestWriteComponent_M2AppTSX_HelloWorld(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("write-component")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-4", Labels: []string{"milestone:m2"}},
		Frontmatter{Files: []string{"src/App.tsx"}, Extras: map[string]string{}},
	)
	if err != nil {
		t.Fatalf("write-component: %v", err)
	}
	body, err := os.ReadFile(filepath.Join(dir, "src/App.tsx"))
	if err != nil {
		t.Fatalf("read: %v", err)
	}
	s := string(body)
	if !strings.Contains(s, `data-testid="app-root"`) {
		t.Error("missing app-root testid")
	}
	if !strings.Contains(s, "Hello world") {
		t.Error("missing Hello world")
	}
	if strings.Contains(s, "TemperatureTable") {
		t.Error("M2 App.tsx should not reference TemperatureTable")
	}
}

func TestWriteComponent_M3AppTSX_TemperatureTableWrapper(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("write-component")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-11", Labels: []string{"milestone:m3"}},
		Frontmatter{Files: []string{"src/App.tsx"}, Extras: map[string]string{}},
	)
	if err != nil {
		t.Fatalf("write-component: %v", err)
	}
	body, _ := os.ReadFile(filepath.Join(dir, "src/App.tsx"))
	s := string(body)
	if !strings.Contains(s, "TemperatureTable") {
		t.Error("M3 App.tsx must wrap TemperatureTable")
	}
	if !strings.Contains(s, `data-testid="app-root"`) {
		t.Error("M3 App.tsx must keep app-root testid")
	}
}

func TestTemperatureTable_TestidsAndStructure(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("write-component")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-9", Labels: []string{"milestone:m3"}},
		Frontmatter{Files: []string{"src/TemperatureTable.tsx"}, Extras: map[string]string{}},
	)
	if err != nil {
		t.Fatalf("write-component: %v", err)
	}
	body, _ := os.ReadFile(filepath.Join(dir, "src/TemperatureTable.tsx"))
	s := string(body)
	if !strings.Contains(s, `data-testid="temperature-table"`) {
		t.Error("must have temperature-table testid")
	}
	if !strings.Contains(s, "row-${r.celsius}") {
		t.Error("must template row-{c} testids")
	}
}

func TestTemperatureRows_FormulaAndRange(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("write-component")
	err := h(
		TemplateContext{ProjectDir: dir, BeadID: "tspa-8", Labels: []string{"milestone:m3"}},
		Frontmatter{Files: []string{"src/temperatureRows.ts"}, Extras: map[string]string{}},
	)
	if err != nil {
		t.Fatalf("write-component: %v", err)
	}
	body, _ := os.ReadFile(filepath.Join(dir, "src/temperatureRows.ts"))
	s := string(body)
	if !strings.Contains(s, "c <= 300") {
		t.Error("must iterate up to 300")
	}
	if !strings.Contains(s, "c += 20") {
		t.Error("must step by 20")
	}
	if !strings.Contains(s, "9") || !strings.Contains(s, "5") || !strings.Contains(s, "32") {
		t.Error("must include 9/5/32 formula constants")
	}
	if !strings.Contains(s, "toFixed(1)") {
		t.Error("must format to 1 decimal place")
	}
}

func TestErrorThenRecover_FirstAttemptFails(t *testing.T) {
	dir := tempDir(t)
	h := GetTemplate("error-then-recover")
	ctx := TemplateContext{ProjectDir: dir, BeadID: "tspa-err-test-1", Labels: []string{}}
	fm := Frontmatter{Extras: map[string]string{}}
	if err := h(ctx, fm); err == nil {
		t.Fatal("first attempt must fail")
	} else if !strings.Contains(err.Error(), "deliberate first-attempt failure") {
		t.Fatalf("wrong error: %v", err)
	}
	if err := h(ctx, fm); err != nil {
		t.Fatalf("second attempt must succeed: %v", err)
	}
}
